core/src/test/java/org/infinispan/expiration/impl/ClusterExpirationFunctionalTest.java

```diff
@@ -50,17 +54,37 @@
 
    protected ConfigurationBuilder configurationBuilder;
 
+   protected StorageType storageType;
+
+   protected ClusterExpirationFunctionalTest storageType(StorageType storageType) {
+      this.storageType = storageType;
+      return this;
+   }
+
+   @Override
+   protected String[] parameterNames() {
+      return concat(super.parameterNames(), "storageType");
+   }
+
+   @Override
+   protected Object[] parameterValues() {
+      return concat(super.parameterValues(), storageType);
+   }
+
    @Override
    public Object[] factory() {
-      return new Object[] {
-            new ClusterExpirationFunctionalTest().cacheMode(CacheMode.DIST_SYNC).transactional(true).lockingMode(LockingMode.OPTIMISTIC),
-            new ClusterExpirationFunctionalTest().cacheMode(CacheMode.DIST_SYNC).transactional(true).lockingMode(LockingMode.PESSIMISTIC),
-            new ClusterExpirationFunctionalTest().cacheMode(CacheMode.DIST_SYNC).transactional(false),
-            new ClusterExpirationFunctionalTest().cacheMode(CacheMode.REPL_SYNC).transactional(true).lockingMode(LockingMode.OPTIMISTIC),
-            new ClusterExpirationFunctionalTest().cacheMode(CacheMode.REPL_SYNC).transactional(true).lockingMode(LockingMode.PESSIMISTIC),
-            new ClusterExpirationFunctionalTest().cacheMode(CacheMode.REPL_SYNC).transactional(false),
-            new ClusterExpirationFunctionalTest().cacheMode(CacheMode.SCATTERED_SYNC).transactional(false),
-      };
+      return Arrays.stream(StorageType.values())
```

Comment:
-1, this is an eviction test and not a storage test.

Reply:
Well it is an expiration test, which can work differently depending upon the storage type.

Reply:
Different how? What's the difference between BINARY and OFF_HEAP, or between BINARY w/ DIST_SYNC and BINARY w/ REPL_SYNC, and do we need to test all locking modes for OBJECT, BINARY and OFF_HEAP?

Reply:
Well they work different for OFF_HEAP as it wasn't even working...

I find that each of the locking modes has different ways of operating as well, I found that optimistic especially can work substantially different and has different intricacies with the storage.

I agree that DIST_SYNC and REPL_SYNC in general shouldn't behave much different, so if we really wanted to we could remove those, but the different locking semantics are very important to test from what I have found.

Reply:
> Well they work different for OFF_HEAP as it wasn't even working...

Sure, there are *some* differences between BINARY and OFF_HEAP, but I don't understand how those differences could affect one of `testLifespanExpiredOnPrimaryOwner` vs. `testLifespanExpiredOnBackupOwner` vs. `testLifespanExpiredOnBoth` without affecting all of them.

And of course there are differences between PESSIMISTIC and OPTIMISTIC locking, but are there any differences in PESSIMISTIC+BINARY vs. PESSIMISTIC+OFF_HEAP?

> I agree that DIST_SYNC and REPL_SYNC in general shouldn't behave much different

Again, I don't mind running *some* tests with REPL_SYNC as well, to show that the repl get optimization doesn't interfere with expiration. But I do mind running *all* the tests with all the storage types and all the locking modes, when the repl get optimization affects all of them in the same way.

Reply:
> Sure, there are some differences between BINARY and OFF_HEAP, but I don't understand how those differences could affect one of testLifespanExpiredOnPrimaryOwner vs. testLifespanExpiredOnBackupOwner vs. testLifespanExpiredOnBoth without affecting all of them.

I have had tests where only a subset have failed.

> And of course there are differences between PESSIMISTIC and OPTIMISTIC locking, but are there any differences in PESSIMISTIC+BINARY vs. PESSIMISTIC+OFF_HEAP?

I don't know, that is the beauty of adding a test for is that I know it is working for that scenario.

> Again, I don't mind running some tests with REPL_SYNC as well, to show that the repl get optimization doesn't interfere with expiration. 

Unfortunately running a subset is not as simple as just adding one line. Do we want to go back to having different classes for the tests? Or do you want me to add a cache mode check at the beginning of each test we don't want to run (we still have startup time)?

>But I do mind running all the tests with all the storage types and all the locking modes, when the repl get optimization affects all of them in the same way.

It doesn't, REPL read optimization does not work with transactions.

> Perhaps you could split the tests in more classes, e.g. one with the tests that are affected by cache mode only, one with the tests that are affected by storage, and one with the tests that are affected by locking?

This sounds like a step back to what we were doing before... which I am not a fan of and I thought were going away from. And tbh cherry picking tests that we *think* are only affected by a specific configuration sounds like a recipe for disaster. We could be wrong and there is nothing guaranteeing that assertion is true in the future (in which case we probably think we have coverage when we may not).

Reply:
> I have had tests where only a subset have failed.

That sounds to me like some `DataContainer` implementations are missing unit tests.

Expiration should be able to rely on `DataContainer` working, and expiration tests should not have to be kept in sync with `DataContainerFactory.construct()` in order to test all the code paths in all the `DataContainer` implementations.

> I don't know, that is the beauty of adding a test for is that I know it is working for that scenario.

You have to stop at some point, if you want the test suite to finish in a reasonable time. In fact, you already did:

* `ClusterExpirationLoaderFunctionalTest` only tests a small subset, how did you decide which combos to test with a loader? Why only with a `DummyInMemoryStore` and not with a `SingleFileStore` as well? Why are eviction and passivation always disabled?
* Even here, you're not testing a cache with eviction, and you're not testing DIST_ASYNC, REPL_ASYNC, INVALIDATION_SYNC, INVALIDATION_ASYNC. No `useSynchronization(true)`, no lock striping, no L1, no encoding, no bias for scattered etc.
* You have `testMaxIdleNodeDies`, but not `testLifespanNodeDies`. You test the node dying just before expiration, but you don't test it dying *during* expiration.

> Unfortunately running a subset is not as simple as just adding one line. Do we want to go back to having different classes for the tests? 

I never said it would be easy, I know it requires thinking hard about what features are really interlocked with each other and what features are (or should be) orthogonal. 

I don't want to go back to one class per cache mode either, because that style had the same problem: it's always easier to add a configuration and run all the existing tests, without considering whether how much that affects the test suite duration and whether they bring any value.

But I believe all the test methods in this class do not depend on the storage type, and that the DataContainer implementations should have their own tests that don't involve expiration or locking. In general, I believe it's impossible to test all the features with all possible configurations, so we must strive to test them independently as much as possible.

Reply:
> > I have had tests where only a subset have failed.
> 
> That sounds to me like some `DataContainer` implementations are missing unit tests.
> 
> Expiration should be able to rely on `DataContainer` working, and expiration tests should not have to be kept in sync with `DataContainerFactory.construct()` in order to test all the code paths in all the `DataContainer` implementations.
> 
> > I don't know, that is the beauty of adding a test for is that I know it is working for that scenario.
> 
> You have to stop at some point, if you want the test suite to finish in a reasonable time. In fact, you already did:
> 
> * `ClusterExpirationLoaderFunctionalTest` only tests a small subset, how did you decide which combos to test with a loader? Why only with a `DummyInMemoryStore` and not with a `SingleFileStore` as well? Why are eviction and passivation always disabled?
> * Even here, you're not testing a cache with eviction, and you're not testing DIST_ASYNC, REPL_ASYNC, INVALIDATION_SYNC, INVALIDATION_ASYNC. No `useSynchronization(true)`, no lock striping, no L1, no encoding, no bias for scattered etc.
> * You have `testMaxIdleNodeDies`, but not `testLifespanNodeDies`. You test the node dying just before expiration, but you don't test it dying _during_ expiration.
> 
> > Unfortunately running a subset is not as simple as just adding one line. Do we want to go back to having different classes for the tests?
> 
> I never said it would be easy, I know it requires thinking hard about what features are really interlocked with each other and what features are (or should be) orthogonal.
> 
> I don't want to go back to one class per cache mode either, because that style had the same problem: it's always easier to add a configuration and run all the existing tests, without considering whether how much that affects the test suite duration and whether they bring any value.
> 
> But I believe all the test methods in this class do not depend on the storage type, and that the DataContainer implementations should have their own tests that don't involve expiration or locking. In general, I believe it's impossible to test all the features with all possible configurations, so we must strive to test them independently as much as possible.

Well I am glad you have infinite time to figure out every possible variant and isolate them to their own test classes. I however do not and am not going to change the tests here. I have been burned way too many times by not having more combinations, not less (and I am pretty sure everyone else can probably agree with that). If you want you can make the changes to the tests, but until then we won't have max idle working for off heap and it won't be making it into 9.4.x anytime soon I guess.

Reply:
And no offense but stuff like this is one of the reasons why we have PRs sitting in review purgatory for so long. This is a working fix and implementation (resolves other issues on CI btw) and tests cases. If you have such a strong objection I would say you should just create a separate JIRA and integrate this. Going back and forth on something that is working, especially the actual implementation, and complaint about how the tests should be reduced is a bit beyond me.

The test now takes 16s instead of 11s according to the difference on master and no difference in total runtime of the test suite.

Reply:
> Even here, you're not testing a cache with eviction, and you're not testing DIST_ASYNC, REPL_ASYNC, INVALIDATION_SYNC, INVALIDATION_ASYNC. No useSynchronization(true), no lock striping, no L1, no encoding, no bias for scattered etc.

1. ASYNC would not work in such a test as these rely on synchronous behavior. 
2. Invalidation doesn't work here either as the ownership doesn't make as much sense, tbh max idle probably doesn't work at all with INVALIDATION :) 
3. The useSynchronization case could probably be added here, sure.
4. Lock striping is that even a supported feature? I thought we were getting rid of it.
5. Does anyone even use L1 nowadays? I feel like it is another one of those features we should kill.
6. Encoding, we technically cover parts of that with OFF_HEAP and BINARY. But sure we can add that if you want.
7. It should be pretty simple to add another option for bias for scattered cache too (just 1 line of code)

Either way one of the disconnects it that you and I see StorageType a bit differently. StorageType is so closely intertwined with all the other features, it isn't just a DataContainer test. Many features work completely different with BINARY as they do OFF_HEAP and even more so with OBJECT. For example Query has specific logic for them (which is another PR that you have shut down and I don't want to work on because of the exact same reason of nitpicking tests). Especially as we move forward I feel that OFF_HEAP will become more and more and more and more prominent and in fact it already is in some cases (e.g. openshift). And adding its support to the functional tests seems exactly what it needs. Unless you somehow think that testing OFF_HEAP with expiration in a LOCAL cache will catch all the issues?

Reply:
> The test now takes 16s instead of 11s according to the difference on master and no difference in total runtime of the test suite.

Ok, I know it's my fault I run with trace logging enabled, but with the latest changes it takes 50s for me (it was 75s with REPL_SYNC).

> StorageType is so closely intertwined with all the other features, it isn't just a DataContainer test. Many features work completely different with BINARY as they do OFF_HEAP and even more so with OBJECT.

That pretty much sums up my concerns. What's so special about StorageType that we can't isolate it and we have to add new tests to every other feature for every StorageType?

> For example Query has specific logic for them (which is another PR that you have shut down and I don't want to work on because of the exact same reason of nitpicking tests).

No, query does not have specific logic for other storage types. Query relies on a feature of core, `WrappedBytes`, which was implemented incorrectly (`instanceof WrappedBytesArray` instead of `instanceof WrappedBytes`). There is still no test for `WrappedBytes` in core, so we can't know whether that feature works without running the query tests.

> It should be pretty simple to add another option for bias for scattered cache too (just 1 line of code)

I guess my comment backfired... I was trying to prove to you that you can never run all possible configs, instead you have to exercise your judgement to decide what to test.

I'm not going to delay integrating this PR any more, but I really feel that StorageTypes shouldn't hide behind other features any more: we should properly define an "internal API" for container storage, and we should have tests exercising this internal API instead of only tests exercising OFF_HEAP+something or BINARY+something.